File: macos/Sources/LaTeXColabCore/WordDiff.swift
import Foundation

public enum DiffKind: Equatable {
    case equal, inserted, deleted
}

public struct DiffSegment: Equatable {
    public let text: String
    public let kind: DiffKind
    public let isWord: Bool
}

struct DiffToken: Equatable {
    let text: String
    let key: String
    let isWord: Bool
}

/// Word-level diff (LCS) used to highlight a rewrite and to count how many
/// words changed.
public enum WordDiff {
    static func tokenize(_ s: String) -> [DiffToken] {
        var out: [DiffToken] = []
        let scalars = Array(s.unicodeScalars)
        var i = 0
        func isWordScalar(_ u: Unicode.Scalar) -> Bool {
            CharacterSet.alphanumerics.contains(u) || u == "'" || u == "\u{2019}"
        }
        while i < scalars.count {
            let u = scalars[i]
            if isWordScalar(u) {
                var j = i
                while j < scalars.count && (isWordScalar(scalars[j]) || (scalars[j] == "-" && j + 1 < scalars.count && isWordScalar(scalars[j + 1]))) { j += 1 }
                let t = String(String.UnicodeScalarView(scalars[i..<j]))
                out.append(DiffToken(text: t, key: t, isWord: true))
                i = j
            } else if u == "\\" {
                var j = i + 1
                while j < scalars.count && CharacterSet.letters.contains(scalars[j]) { j += 1 }
                if j == i + 1 && j < scalars.count { j += 1 } // \% \, \\ etc.
                let t = String(String.UnicodeScalarView(scalars[i..<j]))
                out.append(DiffToken(text: t, key: t, isWord: false))
                i = j
            } else if CharacterSet.whitespacesAndNewlines.contains(u) {
                var j = i
                while j < scalars.count && CharacterSet.whitespacesAndNewlines.contains(scalars[j]) { j += 1 }
                let t = String(String.UnicodeScalarView(scalars[i..<j]))
                out.append(DiffToken(text: t, key: " ", isWord: false))
                i = j
            } else {
                let t = String(u)
                out.append(DiffToken(text: t, key: t, isWord: false))
                i += 1
            }
        }
        return out
    }

    public static func diff(old: String, new: String) -> [DiffSegment] {
        let a = tokenize(old)
        let b = tokenize(new)
        let n = a.count, m = b.count
        if n == 0 { return b.map { DiffSegment(text: $0.text, kind: .inserted, isWord: $0.isWord) } }
        if m == 0 { return a.map { DiffSegment(text: $0.text, kind: .deleted, isWord: $0.isWord) } }

        // Trim the common prefix/suffix first: keeps the DP tiny for small edits.
        var start = 0
        while start < n && start < m && a[start].key == b[start].key { start += 1 }
        var endA = n, endB = m
        while endA > start && endB > start && a[endA - 1].key == b[endB - 1].key { endA -= 1; endB -= 1 }

        var segs: [DiffSegment] = []
        for t in a[0..<start] { segs.append(DiffSegment(text: t.text, kind: .equal, isWord: t.isWord)) }

        let ma = Array(a[start..<endA]), mb = Array(b[start..<endB])
        let na = ma.count, nb = mb.count
        if na > 0 || nb > 0 {
            if na * nb > 6_000_000 {
                for t in ma { segs.append(DiffSegment(text: t.text, kind: .deleted, isWord: t.isWord)) }
                for t in mb { segs.append(DiffSegment(text: t.text, kind: .inserted, isWord: t.isWord)) }
            } else {
                // LCS table, then backtrack.
                let w = nb + 1
                var table = [Int32](repeating: 0, count: (na + 1) * w)
                for i in stride(from: na - 1, through: 0, by: -1) {
                    for j in stride(from: nb - 1, through: 0, by: -1) {
                        if ma[i].key == mb[j].key {
                            table[i * w + j] = table[(i + 1) * w + j + 1] + 1
                        } else {
                            table[i * w + j] = max(table[(i + 1) * w + j], table[i * w + j + 1])
                        }
                    }
                }
                var i = 0, j = 0
                while i < na && j < nb {
                    if ma[i].key == mb[j].key {
                        segs.append(DiffSegment(text: mb[j].text, kind: .equal, isWord: mb[j].isWord))
                        i += 1; j += 1
                    } else if table[(i + 1) * w + j] >= table[i * w + j + 1] {
                        segs.append(DiffSegment(text: ma[i].text, kind: .deleted, isWord: ma[i].isWord))
                        i += 1
                    } else {
                        segs.append(DiffSegment(text: mb[j].text, kind: .inserted, isWord: mb[j].isWord))
                        j += 1
                    }
                }
                while i < na { segs.append(DiffSegment(text: ma[i].text, kind: .deleted, isWord: ma[i].isWord)); i += 1 }
                while j < nb { segs.append(DiffSegment(text: mb[j].text, kind: .inserted, isWord: mb[j].isWord)); j += 1 }
            }
        }
        for t in b[endB..<m] { segs.append(DiffSegment(text: t.text, kind: .equal, isWord: t.isWord)) }
        return segs
    }

    /// Words changed = max(inserted words, deleted words) — a substitution counts once.
    public static func changedWordCount(_ segments: [DiffSegment]) -> Int {
        let ins = segments.filter { $0.kind == .inserted && $0.isWord }.count
        let del = segments.filter { $0.kind == .deleted && $0.isWord }.count
        return max(ins, del)
    }

    /// UTF-16 ranges of changed tokens on each side, for highlighting.
    public static func highlightRanges(_ segments: [DiffSegment]) -> (old: [NSRange], new: [NSRange]) {
        var oldRanges: [NSRange] = []
        var newRanges: [NSRange] = []
        var oldPos = 0, newPos = 0
        for s in segments {
            let len = s.text.utf16.count
            switch s.kind {
            case .equal:
                oldPos += len; newPos += len
            case .deleted:
                oldRanges.append(NSRange(location: oldPos, length: len)); oldPos += len
            case .inserted:
                newRanges.append(NSRange(location: newPos, length: len)); newPos += len
            }
        }
        return (oldRanges, newRanges)
    }
}

File: linux/latexcolab/core/worddiff.py
"""Word-level diff (LCS) used to highlight a rewrite and to count changed words.

Port of WordDiff.swift. Ranges are ``(location, length)`` pairs in Python
string offsets (the Swift original uses UTF-16 offsets; the two agree for all
text a LaTeX paragraph realistically holds and are self-consistent either way).
"""

from __future__ import annotations

import unicodedata
from dataclasses import dataclass
from enum import Enum


class DiffKind(Enum):
    EQUAL = "equal"
    INSERTED = "inserted"
    DELETED = "deleted"


@dataclass(frozen=True)
class DiffSegment:
    text: str
    kind: DiffKind
    is_word: bool
    #: For EQUAL segments: the token as written on the old side (may differ in whitespace).
    original_text: str = ""

    def __post_init__(self) -> None:
        if not self.original_text:
            object.__setattr__(self, "original_text", self.text)


@dataclass(frozen=True)
class _Token:
    text: str
    key: str
    is_word: bool


def _is_word_char(c: str) -> bool:
    return c.isalnum() or c in ("'", "’")


def _is_letter(c: str) -> bool:
    return unicodedata.category(c).startswith("L")


def tokenize(s: str) -> list[_Token]:
    out: list[_Token] = []
    i = 0
    n = len(s)
    while i < n:
        c = s[i]
        if _is_word_char(c):
            j = i
            while j < n and (_is_word_char(s[j])
                             or (s[j] == "-" and j + 1 < n and _is_word_char(s[j + 1]))):
                j += 1
            t = s[i:j]
            out.append(_Token(t, t, True))
            i = j
        elif c == "\\":
            j = i + 1
            while j < n and _is_letter(s[j]):
                j += 1
            if j == i + 1 and j < n:
                j += 1  # \% \, \\ etc.
            t = s[i:j]
            out.append(_Token(t, t, False))
            i = j
        elif c.isspace():
            j = i
            while j < n and s[j].isspace():
                j += 1
            out.append(_Token(s[i:j], " ", False))
            i = j
        else:
            out.append(_Token(c, c, False))
            i += 1
    return out


def diff(old: str, new: str) -> list[DiffSegment]:
    a = tokenize(old)
    b = tokenize(new)
    n, m = len(a), len(b)
    if n == 0:
        return [DiffSegment(t.text, DiffKind.INSERTED, t.is_word) for t in b]
    if m == 0:
        return [DiffSegment(t.text, DiffKind.DELETED, t.is_word) for t in a]

    # Trim the common prefix/suffix first: keeps the DP tiny for small edits.
    start = 0
    while start < n and start < m and a[start].key == b[start].key:
        start += 1
    end_a, end_b = n, m
    while end_a > start and end_b > start and a[end_a - 1].key == b[end_b - 1].key:
        end_a -= 1
        end_b -= 1

    segs: list[DiffSegment] = []
    for i in range(start):
        segs.append(DiffSegment(b[i].text, DiffKind.EQUAL, b[i].is_word, a[i].text))

    ma, mb = a[start:end_a], b[start:end_b]
    na, nb = len(ma), len(mb)
    if na or nb:
        if na * nb > 6_000_000:
            segs.extend(DiffSegment(t.text, DiffKind.DELETED, t.is_word) for t in ma)
            segs.extend(DiffSegment(t.text, DiffKind.INSERTED, t.is_word) for t in mb)
        else:
            # Weighted LCS (matching a word or punctuation mark is worth more
            # than matching a space, so words never look "moved" just to line up
            # whitespace), then backtrack.
            def weight(t: _Token) -> int:
                return 1 if t.key == " " else 3

            w = nb + 1
            table = [0] * ((na + 1) * w)
            for i in range(na - 1, -1, -1):
                row = i * w
                nxt = (i + 1) * w
                ai_key = ma[i].key
                ai_weight = weight(ma[i])
                for j in range(nb - 1, -1, -1):
                    skip = table[nxt + j]
                    right = table[row + j + 1]
                    if right > skip:
                        skip = right
                    if ai_key == mb[j].key:
                        cand = table[nxt + j + 1] + ai_weight
                        table[row + j] = cand if cand > skip else skip
                    else:
                        table[row + j] = skip
            i = j = 0
            while i < na and j < nb:
                if (ma[i].key == mb[j].key
                        and table[i * w + j] == table[(i + 1) * w + j + 1] + weight(ma[i])):
                    segs.append(DiffSegment(mb[j].text, DiffKind.EQUAL, mb[j].is_word, ma[i].text))
                    i += 1
                    j += 1
                elif table[(i + 1) * w + j] >= table[i * w + j + 1]:
                    segs.append(DiffSegment(ma[i].text, DiffKind.DELETED, ma[i].is_word))
                    i += 1
                else:
                    segs.append(DiffSegment(mb[j].text, DiffKind.INSERTED, mb[j].is_word))
                    j += 1
            while i < na:
                segs.append(DiffSegment(ma[i].text, DiffKind.DELETED, ma[i].is_word))
                i += 1
            while j < nb:
                segs.append(DiffSegment(mb[j].text, DiffKind.INSERTED, mb[j].is_word))
                j += 1

    for k in range(m - end_b):
        segs.append(DiffSegment(b[end_b + k].text, DiffKind.EQUAL,
                                b[end_b + k].is_word, a[end_a + k].text))
    return segs


def changed_word_count(segments: list[DiffSegment]) -> int:
    """A substitution counts once: max(inserted words, deleted words)."""
    ins = sum(1 for s in segments if s.kind is DiffKind.INSERTED and s.is_word)
    dele = sum(1 for s in segments if s.kind is DiffKind.DELETED and s.is_word)
    return max(ins, dele)


def highlight_ranges(segments: list[DiffSegment]) -> tuple[list[tuple[int, int]], list[tuple[int, int]]]:
    """``(old, new)`` lists of ``(location, length)`` ranges of changed tokens."""
    old_ranges: list[tuple[int, int]] = []
    new_ranges: list[tuple[int, int]] = []
    old_pos = new_pos = 0
    for s in segments:
        length = len(s.text)
        if s.kind is DiffKind.EQUAL:
            old_pos += len(s.original_text)
            new_pos += length
        elif s.kind is DiffKind.DELETED:
            old_ranges.append((old_pos, length))
            old_pos += length
        else:
            new_ranges.append((new_pos, length))
            new_pos += length
    return old_ranges, new_ranges
